theano/gof/utils.py

```diff
@@ -217,14 +217,21 @@ def __call__(cls, *args, **kwargs):
         props = cls.__dict__.get('__props__', None)
         if props is None:
             return super(MetaObject, cls).__call__(*args, **kwargs)
+
         key = [cls]
         for arg in args:
             arg = make_hashable(arg)
             key.append(arg)
+            arg_dtype = get_arg_dtype(arg)
+            if arg_dtype:
+                key.extend(arg_dtype)
         for k, v in kwargs.items():
             key.append(k)
             v = make_hashable(v)
             key.append(v)
+            v_dtype = get_arg_dtype(v)
+            if v_dtype:
+                key.extend(v_dtype)
```

Comment:
>  If one of the kwargs key isn't a prop, we could skip the cache.

This was exactly my point for implementation. But from what I understand, @ReyhaneAskari in some cases, is having an Elemwise with the elements of props being the same. I have an idea, I will comment at the appropriate place.